### ISSUE TITLE
2-dimensional KDE outputs skewed results

### ISSUE THREAD

**Issue by waeiski** [mwaskom/seaborn#3472]
There appears to be a potential problem in Seaborn's way of doing 2-dimensional KDE with `scipy.stats.gaussian_kde`. It appears that either `seaborn` or `scipy` is calculating the extent of the plot separately for X and Y coordinates, which causes the output to be skewed or elongated on the "longer" axis. I believe this is not expected, nor desired behaviour and thus can potentially affect publications, analyses etc.

To demonstrate, I will use some point data from Portugal as it is a country that is long on the north-south axis and rather thin on the east-west axis. First, I will run the 2-dimensional KDE using code for `seaborn` and provide the "skewed" output. Second, I will repeat the analysis using `matplotlib` and `sklearn.neighbors.KernelDensity` and provide the expected output.

Here is the `seaborn` example with the resulting skewed plot.
```
import geopandas as gpd
import seaborn as sns
import matplotlib.pyplot as plt
import contextily

# read data
df = gpd.read_file('point_data.gpkg')

# plot seaborn kde
fig, ax = plt.subplots(figsize=(15,15))
df.plot(ax=ax, markersize=0.1, color='k', zorder=2)
g = sns.kdeplot(data=df, x=df['geometry'].x, y=df['geometry'].y,
                fill=True, cmap='crest', alpha=.6, levels=7)

# add basemap
contextily.add_basemap(
    ax=ax,
    crs=df.crs,
    source="CartoDB Positron No Labels")
```
![seaborn_example](https://github.com/mwaskom/seaborn/assets/31708782/51bb9dc3-a397-4b81-a838-fb133369fe2c)

Here is the code to replicate the expected result using `contourf` from `matplotlib` on a gaussian KDE calculated by `KernelDensity` from `sklearn.neighbors`:
```
import geopandas as gpd
import seaborn as sns
import matplotlib.pyplot as plt
import contextily
import numpy as np
from sklearn.neighbors import KernelDensity

# read data
df = gpd.read_file('point_data.gpkg')


# instantiate sklearn KDE
kde = KernelDensity(bandwidth=15000, kernel='gaussian')
kde.fit(df.get_coordinates().values)

# get bounds of data
bounds = df.total_bounds

# create mesh for predicting KDE on with more space around the points
x_mesh, y_mesh = np.meshgrid(np.arange(bounds[0] - 35000, bounds[2] + 35000, 3000),
                             np.arange(bounds[1] - 35000, bounds[3] + 35000, 3000))

# get the prediction
pred = kde.score_samples(np.vstack([x_mesh.flatten(), y_mesh.flatten()]).T)

# set levels for countour plots
levels = np.linspace(-30, pred.max(), 7)

# plot
ax = df.plot(zorder=2, markersize=0.1, figsize=(15,15), color='k')
plt.contourf(x_mesh, y_mesh, pred.reshape(x_mesh.shape), levels=levels,
             cmap='crest', alpha=.6)
contextily.add_basemap(
    ax=ax,
    crs=df.crs,
    source="CartoDB Positron No Labels")
```
![sklearn_example](https://github.com/mwaskom/seaborn/assets/31708782/c31cfb02-c64b-4de8-9e65-55905ddfb8ac)

This might not be caused by anything `seaborn` does, but by `scipy`. Either way, `seaborn` is affected and, at least to me, this deserved to be brought up as a potential issue. Unfortunately, I lack the time and skills to effectively find the line of code causing this, so I can't verify where this "skewing" occurs. 

My library versions:
`seaborn 0.12.2`
`matplotlib 3.7.1`
`geopandas 0.13.2`
`contextily 1.3.0`
`numpy 1.24.4`
`sklearn 1.3.0`

p.s. I want to thank Martin Fleischmann for helping with the troubleshooting.

**Comment by mwaskom** [comment]
This just looks to me like different bandwidth parameterizations?

**Comment by martinfleis** [comment]
It looks that scipy defines different bandwidths for each axis, while sklearn uses one along both, which is what you'd expect for a use case like this. I suppose that any KDE is skewed, it is just not noticeable as long as the extent is roughly the same along both dimensions. One you get elongated data like those above, it gets more pronounced.

**Comment by mwaskom** [comment]
I think that's right. But I also think it's just a tradeoff — a single bandwidth parameter does not make sense when your dimensions have different natural scales. (For scikit learn's ML usecases, you'd often have an upstream normalization step).

In any case that's the way the scipy parameterization works and seaborn's not going to add a scikit-learn dependency just because of this, so I'm not sure what the seaborn ask is here?

**Comment by martinfleis** [comment]
Yeah, for the use case in seaborn, where kde is used to visualise data from two independent variables using different scales, the behavior of scipy is the one you want. It just doesn't work when variables represent location of an observation on a plane, like with this example of geospatial data. The only case when scipy's algorithm would work would be if the bounding box of the dataframe was a perfect square. Otherwise, there is a distortion and more elongated the bounding box is, more pronounced this distortion is.

Unless scipy adds a support of a fixed bandwidth, I guess there's not much seaborn can do. I'll open an issue there.

The only thing I can think of is adding a note to the kdeplot docstring clarifying this behaviour and making it clear that it is not fit for a use case like this.

**Comment by mwaskom** [comment]
IMO you're overstating things to say it is "not suitable" generally speaking but in any case I think I'll close here for now.

**Comment by mdhaber** [comment]
>In any case that's the way the scipy parameterization works 

@mwaskom Would you be interested in supporting this  if SciPy were to accept an array to be used as the data covariance matrix? https://github.com/scipy/scipy/pull/19297 would implement this. I'm not sure how seaborn would want to support the user's request, but one possibility would be to accept `data_covariance` as an optional parameter and pass it along to `scipy.stats.kde`.

**Comment by mwaskom** [comment]
I don't think I understand how this would be used. The parameter description in the linked PR says "If not provided, this will be computed from `dataset`". When would you pass a covariance matrix that does not correspond to the sample covariance of your data?

Note that seaborn contains a lightweight vendored version of the scipy kde code (so that it does not require scipy for installation; although it does call into scipy instead if scipy is installed to pick up their Cython implementation) so supporting additional kde options could be a headache.

**Comment by mwaskom** [comment]
OK I read up the linked issue chain and I think the idea is that you would enable this use-case by passing an identity matrix to enable a fixed bandwidth. I left a comment on the scipy PR about how this feels unintuitive — AFAICT it's basically supporting this use-case by opening up a hack. I'm not certain I want this to leak all the way up to seaborn, but could potentially be open to using it to support a fixed bandwidth parameterization. (FWIW it's useful both to be able to use the same bandwidth on multiple dimensions *and* to specify the bandwidth in data units which is what seems to happen when you pass an identity matrix for the data covariance?)
